Sources/Core/Sources/Render/ScreenRenderer.swift

```diff
@@ -0,0 +1,131 @@
+import Foundation
+import MetalKit
+
+/// The renderer managing offscreen rendering and displaying final result in view.
+public final class ScreenRenderer: Renderer {
+  // MARK: - Private properties
+  /// The device used to render.
+  private var device: MTLDevice
+  
+  /// Renderer's own pipeline state (for drawing on-screen)
+  private var pipelineState: MTLRenderPipelineState
+  
+  /// Offscreen render pass used to command rendering into renderer's internal textures
+  private var offscreenRenderPass: MTLRenderPassDescriptor!
+  
+  /// Renderer's profiler
+  private var profiler: Profiler<RenderingMeasurement>
+  
+  /// Render target texture into which offscreen rendering is performed
+  private var renderTargetTexture: MTLTexture!
+  
+  /// Render target depth texture
+  private var renderTargetDepthTexture: MTLTexture!
+  
+  /// Client for which rendering is performed
+  private var client: Client
+  
+  // MARK: - Init
+  public init(
+    client: Client,
+    device: MTLDevice,
+    profiler: Profiler<RenderingMeasurement>
+  ) throws {
+    self.device = device
+    self.client = client
+    self.profiler = profiler
+    
+    // Create pipeline state
+    let library = try MetalUtil.loadDefaultLibrary(device)
+    pipelineState = try MetalUtil.makeRenderPipelineState(
+      device: device,
+      label: "ScreenRenderer",
+      vertexFunction: try MetalUtil.loadFunction("screenVertexFunction", from: library),
+      fragmentFunction: try MetalUtil.loadFunction("screenFragmentFunction", from: library),
+      blendingEnabled: false
+    )
+  }
+  
+  // MARK: - Public
+  public var renderDescriptor: MTLRenderPassDescriptor {
+    return offscreenRenderPass
+  }
+  
+  public var renderOutputTexture: MTLTexture? {
```

Comment:
Why did you make these computed properties instead of just making the underlying properties public? They seem a bit redundant

Reply:
Similarly, these were made with upscaling in mind. I'd have a set of 1x, 2x and 4x combinations for render target and upscaled output textures, using computed properties i'd specify correct behaviour for different upscaling factors. I'll go ahead and remove it at the moment.

Reply:
Ok kinda makes sense, but yeah best to keep it simpler for now